pulsar-client/src/main/java/org/apache/pulsar/client/impl/TransactionMetaStoreHandler.java

```diff
@@ -335,41 +402,77 @@ void handleEndTxnResponse(CommandEndTxnResponse response) {
             }
             op.callback.complete(null);
         } else {
-            LOG.error("Got end txn response for request {} error {}", response.getRequestId(), response.getError());
             handleTransactionFailOp(response.getError(), response.getMessage(), op);
+            if (response.getError() == ServerError.TransactionCoordinatorNotFound) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Get a response for request {} error TransactionCoordinatorNotFound and try it again",
+                            response.getRequestId());
+                }
+                timer.newTimeout(timeout -> {
+                    long requestId = client.newRequestId();
+                    TxnID txnID = op.txnID;
+                    TxnAction action = op.action;
+                    BaseCommand cmd = Commands.newEndTxn(requestId, txnID.getLeastSigBits(),
+                            txnID.getMostSigBits(), action);
+                    OpForEndTxnCallBack opNew = OpForEndTxnCallBack
+                            .create(Commands.serializeWithSize(cmd), op.callback, txnID, action);
+                    op.cmd.release();
+                    op.recycle();
+                    tryExecuteCommandAgain(opNew, requestId);
+                }, op.backoff.next(), TimeUnit.MILLISECONDS);
+                return;
+            } else {
+                LOG.error("Got end txn response for request {} error {}", response.getRequestId(), response.getError());
+            }
         }
-
         onResponse(op);
     }
 
+    private <T> void tryExecuteCommandAgain(OpBase<T> op, long requestId) {
+        if (cnx() == null) {
+            timer.newTimeout(timeout ->
+                    tryExecuteCommandAgain(op, requestId), op.backoff.next(), TimeUnit.MILLISECONDS);
+            return;
+        }
+        pendingRequests.put(requestId, op);
+        timeoutQueue.add(new RequestTime(System.currentTimeMillis(), requestId));
+        cnx().ctx().writeAndFlush(op.cmd, cnx().ctx().voidPromise());
```

Comment:
should use local variable, otherwise it will produce NPE. code like this
```
       ClientCnx cnx = cnx();
        if (cnx == null) {
            timer.newTimeout(timeout ->
                    tryExecuteCommandAgain(op, requestId), op.backoff.next(), TimeUnit.MILLISECONDS);
            return;
        }
        pendingRequests.put(requestId, op);
        timeoutQueue.add(new RequestTime(System.currentTimeMillis(), requestId));
        cnx.ctx().writeAndFlush(op.cmd, cnx().ctx().voidPromise());
```